harness/determined/core/_checkpoint.py

```diff
@@ -241,79 +243,124 @@ def upload(
         self._report_checkpoint(storage_id, resources, metadata)
         return storage_id
 
     def _upload_sharded(
-        self, ckpt_dir: Optional[str], storage_id: str, metadata: Optional[Dict[str, Any]] = None
+        self,
+        ckpt_dir: Optional[str],
+        storage_id: str,
+        metadata: Optional[Dict[str, Any]] = None,
+        *,
+        selector: Optional[Callable[[str], bool]] = None,
     ) -> str:
+
+        if selector is not None and ckpt_dir is None:
+            raise RuntimeError("ckpt_dir has to be provided if selector is not None")
+
         ckpt_dir_mask = self._dist.allgather(ckpt_dir is not None)
         if not any(ckpt_dir_mask):
             raise RuntimeError(
                 "cannot call .upload(ckpt_dir=None, shard=True), from all ranks; "
                 "at least one rank must have a valid ckpt_dir"
             )
 
-        # Deconflict locally-shared directories; if every worker uploads /tmp/ckpt, then only
-        # the lowest rank on each node will actually upload this directory.
+        # Deconflict locally-shared directories; if multiple ranks (with no selectors) upload
+        # /tmp/ckpt then only the lowest rank on each node will actually upload this directory.
+        # If multiple ranks (with selectors) want to upload selected files from /tmp/ckpt, then
+        # each rank will upload its selected files.
         if ckpt_dir is None:
             file_uid = None
         else:
             st = os.stat(ckpt_dir)
             file_uid = (st.st_dev, st.st_ino)
+
         all_file_uids = self._dist.allgather(file_uid)
-        # Decide if our rank is the lowest rank trying to upload this ckpt_dir.
-        want_upload = file_uid and all_file_uids.index(file_uid) == self._dist.rank
 
-        # Decide what we are going to upload.
+        # Rank uploads resources only if (1) it has a selector; or
+        # (2) it is the lowest local rank requested to upload this ckpt_dir.
+        want_upload = selector is not None or (
+            file_uid is not None and all_file_uids.index(file_uid) == self._dist.rank
+        )
+        # Collect and filter resources for all uploading ranks.
         if want_upload:
             assert ckpt_dir
             resources = self._storage_manager._list_directory(ckpt_dir)
+            if selector is not None:
+                resources = {key: resources[key] for key in resources if selector(key)}
+
+            # Metadata.json is a special file that is created and uploaded separately.
+            resources.pop("metadata.json", None)
         else:
             resources = {}
 
-        # Merge resources, detect conflicts.
         all_resources = self._dist.allgather(resources)
         merged_resources, conflicts = merge_resources(all_resources)
         if conflicts:
-            self._try_resolving_conflicts(ckpt_dir, conflicts)
+            self._resolve_conflicts(ckpt_dir, conflicts)
```

Comment:
This is such a tiny nit.  Feel free to ignore it.

It would be really elegant if either `_resolve_conflicts` calculated a new `resources`, like:

```
merged_resources, conflicts = merge_resources(all_resources)
resources = self._resolve_conflicts(conflicts, resources)
```

That way it is obvious `resources` only contains deconflicted paths, and you wouldn't need the long comment or `filter()` call a few lines later.

Reply:
That is an elegant solution! Modified accordingly.